### DISCUSSION TITLE
How to set a dynamic header for specific pages with App Router?

### DISCUSSION THREAD

**Discussion by CaptShmo** [vercel/next.js#49733] [App Router]
Hi, I have a nextjs app which gets embeded on another site through an iframe. In order for the other site to display the iframe, they require that I set the frame-ancestors CSP header to a specific value that they will give me through a fetch request. I think this is going to prevent me from using the next.config.js method of setting headers because I need to do data fetching before I set the header because the value comes from them and is dynamic. Also, because it's an iframe that needs to have that header set, I don't think using the "route.js" method is going to work either, because the result they are expecting is an html page with the proper header set. I suppose I could use route.js to return raw html but that sort of method sort of skips over a lot of the useful features of nextjs. Before I was handling this by using the setHeader() method on the context response in getServerSideProps() and it worked. Is there any corresponding methodology I can use for nextjs server (page) components in App Router? Thanks.

Edit: more elaboration on the exact workflow required here:

1) External site tries to load my nextjs page as an iframe, with a url like https://my-nextjs-app.com/specific-page?value=123
2) My nextjs app needs to do a fetch to https://the-other-site.com/value/123 using the query param they set with the iframe request
3) The fetch request returns data with a value like "dynamic-frame-ancestor = abc"
4) Then I send back my nextjs html result with the frame-ancestor csp header set to "abc". 
5) The other site sees the header includes the dynamic value they sent and renders the iframe. If it's not there or doesn't match, they won't display the iframe.

**Comment by volpster2100** [comment]
Sounds like you need to use middleware?

**Reply by CaptShmo** [reply]
I tried this once before for a similar problem in nextjs12 and couldn't get it to work with their example code, but checking this again looks like they actually have a bug in their documentation for this feature, I went to the source code for NextResponse and it seems like their documentation is out of date. I will file an issue there. Thank you, I will mark this as the answer later if I can get your strategy to work.

**Reply by volpster2100** [reply]
Np! Lmk if you hit any snags. I didn’t end up using middleware because it only runs in the edge so it only supports fetch api but none of the other node apis. Good luck!!

**Comment by k-ruben** [comment]
I have the same Problem. There is no migration guide for use-cases where the `context` was used in `getServerSideProps`.
There are two cases for me: 
1. I need the cookies from the response object that are not yet set on the client
2. I need to set the Cache-Control to public/private depending on a login-state
